standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DirectSqlUpdatePart.java

```diff
@@ -602,18 +594,18 @@ public long getNextCSIdForMPartitionColumnStatistics(long numStats) throws MetaE
       try (Statement statement = dbConn.createStatement()) {
         statement.executeUpdate(query);
       }
-      dbConn.commit();
-      committed = true;
+      closeDbConn(jdoConn);
+      tx.commit();
       return maxCsId;
     } catch (Exception e) {
       LOG.error("Unable to getNextCSIdForMPartitionColumnStatistics", e);
       throw new MetaException("Unable to getNextCSIdForMPartitionColumnStatistics  "
               + " due to: " + e.getMessage());
     } finally {
-      if (!committed) {
-        rollbackDBConn(dbConn);
+      if (tx.isActive()) {
+        closeDbConn(jdoConn);
```

Comment:
should it be
````
closeDbConn(jdoConn);
if (tx.isActive()) {
  tx.rollback();
}
````

Reply:
We have a `closeDbConn(jdoConn)` before tx commits, if the `jdoConn` is being closed twice, a NucleusUserException is thrown from https://github.com/datanucleus/datanucleus-core/blob/master/src/main/java/org/datanucleus/store/NucleusConnectionImpl.java#L54-L62, then we log the exception at`closeDbConn(jdoConn);`, it could be misleading however.

Reply:
can we check in `closeDbConn` if jdoConn is closed and if not attempt to close it?

Reply:
````!dbConn.isClosed()````

Reply:
why not wrap query execution in try-finally and close db connection there?
````
      try (Statement statement = dbConn.createStatement()) {
        statement.executeUpdate(query);
      } finally {
        closeDbConn(jdoConn);
      }
      tx.commit();
````

Reply:
````
 } finally {
            if (tx.isActive()) {
                tx.rollback();
            }
````

Reply:
If the exception throws from https://github.com/apache/hive/blob/c6965423da67515c4fc8ba70a1d10ba15624789d/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DirectSqlUpdatePart.java#L563-L586 , we still need to close the `jdoConn` at the outer finally,

Reply:
ok, back to the prev approach, can't we check if jdoConn is not already closed?
````
closeDbConn(jdoConn); 
if (tx.isActive()) { 
    tx.rollback(); 
}
...
void closeDbConn(JDOConnection jdoConn) {
    try {
      if (dbConn != null && !dbConn.isClosed()) {
        jdoConn.close();
      }
````

Reply:
sounds the `JDOConnection` doesn't provide the method `isClosed()` to check the state...

Reply:
what do you mean? check the rollback method that was removed

Reply:
how about a nested try, since it doesn't implement auto-closeble
````
try {
    jdoConn = pm.getDataStoreConnection();
....
} finally {
    closeDbConn(jdoConn);
}
tx.commit();
````

Reply:
ack, thank you for the suggestion!